cpp/path_module/algorithm/path.hpp

```diff
@@ -44,64 +45,113 @@ struct LabelSets {
+class PathHelper {
+ public:
+  explicit PathHelper(const mgp::List &labels, const mgp::List &relationships, int64_t min_hops, int64_t max_hops);
+  explicit PathHelper(const mgp::Map &config);
 
-bool ShouldExpand(const LabelBools &labelBools, const LabelBoolsStatus &labelStatus);
+  RelDirection GetDirection(const std::string &rel_type) const;
+  LabelBools GetLabelBools(const mgp::Node &node) const;
 
-void Expand(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory);
+  bool AnyDirected(bool outgoing) const { return outgoing ? config_.any_outgoing : config_.any_incoming; }
+  bool IsNotStartOrSupportsStartNode(bool is_start) const { return (config_.filter_start_node || !is_start); }
+  bool IsNotStartOrSupportsStartRel(bool is_start) const { return (config_.begin_sequence_at_start || !is_start); }
 
-void FilterLabel(const std::string_view label, const LabelSets &labelFilters, LabelBools &labelBools);
+  bool AreLabelsValid(const LabelBools &label_bools) const;
+  bool ContinueExpanding(const LabelBools &label_bools, size_t path_size) const;
 
-void ParseLabels(const mgp::List &list_of_labels, LabelSets &labelSets);
+  bool PathSizeOk(int64_t path_size) const;
+  bool PathTooBig(int64_t path_size) const;
+  bool Whitelisted(bool whitelisted) const;
 
-void ParseRelationships(const mgp::List &list_of_relationships, RelationshipSets &relationshipSets, bool &any_outgoing,
-                        bool &any_incoming);
-bool PathSizeOk(const int64_t path_size, const int64_t &max_hops, const int64_t &min_hops);
+  // methods for parsing config
+  void FilterLabelBoolStatus();
+  void FilterLabel(std::string_view label, LabelBools &label_bools) const;
+  void ParseLabels(const mgp::List &list_of_labels);
+  void ParseRelationships(const mgp::List &list_of_relationships);
 
-bool RelationshipAllowed(const std::string &rel_type, const RelationshipSets &relationshipSets, bool &any_outgoing,
-                         bool &any_incoming, bool outgoing);
+ private:
+  Config config_;
+};
 
-bool Whitelisted(const bool &whitelisted, const bool &whitelist_empty);
+class PathExpand {
+ public:
+  explicit PathExpand(PathHelper &&helper, const mgp::RecordFactory &record_factory, const mgp::Graph &graph)
+      : helper_(helper), record_factory_(record_factory), graph_(graph) {}
+
+  void ExpandPath(mgp::Path &path, const mgp::Relationship &relationship, int64_t path_size);
+  void ExpandFromRelationships(mgp::Path &path, mgp::Relationships relationships, bool outgoing, int64_t path_size,
+                               std::set<std::pair<std::string_view, int64_t>> &seen);
+  void StartAlgorithm(mgp::Node node);
+  void Parse(const mgp::Value &value);
+  void DFS(mgp::Path &path, int64_t path_size);
+  void RunAlgorithm();
+
+ private:
+  PathHelper helper_;
+  const mgp::RecordFactory &record_factory_;
+  const mgp::Graph &graph_;
+  std::unordered_set<int64_t> visited_;
+  std::unordered_set<mgp::Node> start_nodes_;
```

Comment:
I remembered, you can make this a class, and make PathSubgraph and PathExpand friends of this class and then you can reference private variables

Reply:
Yes, great suggestion!